File: android/src/main/java/com/nitrogeolocation/LocationUtils.kt
package com.nitrogeolocation

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.os.Build
import android.os.SystemClock
import androidx.core.app.ActivityCompat
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.margelo.nitro.nitrogeolocation.GeoCoordinates
import com.margelo.nitro.nitrogeolocation.GeoPosition

/**
 * Utility functions for location operations
 */
object LocationUtils {
    
    /**
     * Check if Google Play Services is available
     */
    fun isGooglePlayServicesAvailable(context: Context): Boolean {
        val availability = GoogleApiAvailability.getInstance()
        val resultCode = availability.isGooglePlayServicesAvailable(context)
        return resultCode == ConnectionResult.SUCCESS
    }
    
    /**
     * Check if location permission is granted
     */
    fun hasLocationPermission(context: Context): Boolean {
        return ActivityCompat.checkSelfPermission(
            context, 
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED ||
        ActivityCompat.checkSelfPermission(
            context, 
            Manifest.permission.ACCESS_COARSE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
    }
    
    /**
     * Get the age of a location in milliseconds
     */
    fun getLocationAge(location: Location): Long {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            (SystemClock.elapsedRealtimeNanos() - location.elapsedRealtimeNanos) / 1_000_000
        } else {
            System.currentTimeMillis() - location.time
        }
    }
    
    /**
     * Convert Android Location to GeoPosition
     */
    fun locationToGeoPosition(location: Location): GeoPosition {
        val altitudeAccuracy = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            location.verticalAccuracyMeters.toDouble()
        } else {
            null
        }

        // isMock() was added in API 31 (Android S), isFromMockProvider deprecated but still works
        val mocked = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            location.isMock
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
            @Suppress("DEPRECATION")
            location.isFromMockProvider
        } else {
            null
        }

        val coords = GeoCoordinates(
            latitude = location.latitude,
            longitude = location.longitude,
            accuracy = location.accuracy.toDouble(),
            altitude = location.altitude,
            altitudeAccuracy = altitudeAccuracy,
            heading = location.bearing.toDouble(),
            speed = location.speed.toDouble()
        )

        return GeoPosition(
            coords = coords,
            timestamp = location.time.toDouble(),
            provider = location.provider,
            mocked = mocked
        )
    }
}

File: ios/LocationUtils.swift
//
//  LocationUtils.swift
//  NitroGeolocation
//
//  Created by Rom on 11/26/2025.
//

import Foundation
import CoreLocation
import NitroModules

/**
 * Utility functions for location operations.
 */
struct LocationUtils {
    
    /**
     * Check if location services are enabled and authorized.
     *
     * - Returns: true if location services are available
     */
    static func hasLocationPermission() -> Bool {
        guard CLLocationManager.locationServicesEnabled() else {
            return false
        }
        
        let status = getAuthorizationStatus()
        return status == .authorizedWhenInUse || status == .authorizedAlways
    }
    
    /**
     * Get the authorization status for location services.
     * Compatible with iOS 13 and iOS 14+
     *
     * - Returns: The current authorization status
     */
    static func getAuthorizationStatus() -> CLAuthorizationStatus {
        if #available(iOS 14.0, *) {
            return CLLocationManager().authorizationStatus
        } else {
            return CLLocationManager.authorizationStatus()
        }
    }
    
    /**
     * Check if location services are enabled on the device.
     *
     * - Returns: true if location services are enabled
     */
    static func isLocationServicesEnabled() -> Bool {
        return CLLocationManager.locationServicesEnabled()
    }
    
    /**
     * Get the age of a location in milliseconds.
     *
     * - Parameter location: The location to check
     * - Returns: Age in milliseconds
     */
    static func getLocationAge(_ location: CLLocation) -> Double {
        return -location.timestamp.timeIntervalSinceNow * 1000.0
    }
    
    /**
     * Convert a CLLocation to a GeoPosition struct.
     *
     * - Parameter location: The CoreLocation location object
     * - Returns: A GeoPosition struct suitable for returning to JavaScript
     */
    static func locationToGeoPosition(_ location: CLLocation) -> GeoPosition {
        let coords = GeoCoordinates(
            latitude: location.coordinate.latitude,
            longitude: location.coordinate.longitude,
            accuracy: location.horizontalAccuracy,
            altitude: location.altitude,
            altitudeAccuracy: location.verticalAccuracy >= 0 ? location.verticalAccuracy : nil,
            heading: location.course >= 0 ? location.course : nil,
            speed: location.speed >= 0 ? location.speed : nil
        )
        
        let timestamp = location.timestamp.timeIntervalSince1970 * 1000.0 // Convert to milliseconds
        
        return GeoPosition(
            coords: coords,
            timestamp: timestamp,
            provider: "CoreLocation",
            mocked: false
        )
    }
}
